src/south/ocnos/goldstone/south/ocnos/interfaces.py

```diff
@@ -0,0 +1,917 @@
+            except ncclient.operations.RPCError as e:
+                self.server.ocnos_conn.discard_changes()
+                raise sysrepo.SysrepoInternalError(
+                    f"Edit-config failed in OcNOS for set interface-mode: {e.message}"
+                )
+            global GS_IF_MODE
+            GS_IF_MODE[self.ifname] = value.lower()
+            logger.debug(f"set {self.ifname} interface-mode with {value}")
+        else:
+            try:
+                self.unset_interface_mode(user)
+            except ncclient.operations.RPCError as e:
+                self.server.ocnos_conn.discard_changes()
+                raise sysrepo.SysrepoInternalError(
+                    f"Edit-config failed in OcNOS for unset access-vlans: {e.message}"
+                )
+            logger.debug(f"unset {self.ifname} interface-mode")
+
+    def set_interface_mode(self, value):
+        assert value in ["ACCESS", "TRUNK"]
+        value = "access" if value == "ACCESS" else "trunk"
+        # Add /interfaces/interface[name=ifname]/config[name=ifname]/enable-switchport
+        xpath_if_swport = IPI_INTERFACE_CONFIG_SWITCHPORT.format(
+            self.ifname, self.ifname
+        )
+        xml_if_swport = self.server.ocnos_conn.xpath2xml(xpath_if_swport)
+        xpath_if_mode = IPI_PORT_VLAN_SWITCHED_VLAN_CONFIG_INTERFACE_MODE.format(value)
+        xml_if_mode = self.server.ocnos_conn.xpath2xml(xpath_if_mode, value=value)
+        insert_xml(xml_if_mode, xml_if_swport, TAG_IPI_INTERFACE)
+
+        # Add /network-instances/network-instance/bridge-ports/interface[name=ifname]/config[name=ifname]
+        xpath_ni = IPI_NETWORK_INSTANCE.format(
+            INSTANCE_NAME_FOR_IEEE_VLAN_BRIDGE, INSTANCE_TYPE
+        )
+        xml_ni = self.server.ocnos_conn.xpath2xml(xpath_ni)
+        xpath_bridge_ports = IPI_BRIDGE_PORTS_INTERFACE_CONFIG_INTERFACE.format(
+            self.ifname, self.ifname
+        )
+        xml_bridge_ports = self.server.ocnos_conn.xpath2xml(xpath_bridge_ports)
+        insert_xml(xml_bridge_ports, xml_ni, TAG_IPI_NETWORK_INSTANCE)
+
+        config = new_ele("config")
+        config.append(xml_if_swport)
+        config.append(xml_ni)
+        logger.debug(
+            f"Sending xml for interface-mode configuration: {etree.tostring(config,pretty_print=True).decode()}"
+        )
+        return self.server.ocnos_conn.conn.edit_config(config)
+
+    def unset_interface_mode(self, user):
+        running_config = self.server.get_running_data(
+            GOLDSTONE_INTERFACE_TOP,
+            default={},
+            strip=False,
+            include_implicit_defaults=True,
+        )
+        xpath_gs_if_mode = GS_IF_MODE_CONFIG.format(self.ifname)
+        if_mode = libyang.xpath_get(running_config, xpath_gs_if_mode, None)
+        if if_mode is None:
+            if self.type == "deleted":
+                global GS_IF_MODE
+                if_mode = GS_IF_MODE.get(self.ifname)
```

Comment:
What if `interface-mode` is already configured when `south-ocnos` is started, and the configuration is handled in `reconcile()` then the `interface-mode` config is deleted? `GS_IF_MODE` doesn't have any entry, and `if_mode` will become `None` here.

Reply:
Also, `GS_IF_MODE` doesn't need to be a global variable. Please store it as a member of `InterfaceServer` and access it like `self.server.gs_if_mode` in the change handler.

Reply:
> What if `interface-mode` is already configured when `south-ocnos` is started, and the configuration is handled in `reconcile()` then the `interface-mode` config is deleted? `GS_IF_MODE` doesn't have any entry, and `if_mode` will become `None` here.

Current implementation does **NOT** delete `interface-mode` in `reconcile()` in that case.
When we unset `interface-mode` for an interface, we need to send the configured value (`access` or `trunk`) to OcNOS. However, we cannot obtain that value in `ChangeDeleted` callback. It seems that only xpath is obtainable like:
`event: done, changes: [ChangeDeleted(/goldstone-interfaces:interfaces/interface[name='xe1']/goldstone-vlan:switched-vlan/config/interface-mode)]`

Also, it seems that the value has already been deleted from running datastore at that time.
So we cannot get it from the datastore and the value returned [here](https://github.com/oopt-goldstone/goldstone-mgmt/pull/71/files#diff-64dd4da385d6884571543f3fee097fba863d284e8a6dee1ae5f2b7ac2f03a8f6R306-R311) is `None`.
That's why we introduced the conditional branch and global variable of `GS_IF_MODE`.

> Please store it as a member of InterfaceServer and access it like self.server.gs_if_mode in the change handler.

Thank you so much for telling me more clever way. I fixed it.